packages/git/src/browser/git-sync-service.ts

```diff
@@ -0,0 +1,227 @@
+        }
+        const { upstreamBranch } = state;
+        const methods: {
+            label: string
+            warning: string
+            value: GitSyncService.SyncMethod
+        }[] = [{
+            label: `Pull and push commits from and to '${upstreamBranch}'`,
+            warning: `This action will pull and push commits from and to '${upstreamBranch}'.`,
+            value: 'pull-push'
+        }, {
+            label: `Fetch, rebase and push commits from and to '${upstreamBranch}'`,
+            warning: `This action will fetch, rebase and push commits from and to '${upstreamBranch}'.`,
+            value: 'rebase-push'
+        }, {
+            label: `Force push commits to '${upstreamBranch}'`,
+            warning: `This action will override commits in '${upstreamBranch}'.`,
+            value: 'force-push'
+        }];
+        const method = await this.pick(`Pick how changes should be synchronized:`, methods);
+        if (method && await this.confirm('Synchronize Changes', methods.find(({ value }) => value === method)!.warning)) {
+            return method;
+        }
+        return undefined;
+    }
+
+    canPublish(): boolean {
+        if (this.syncing) {
+            return false;
+        }
+        const status = this.repositoryTracker.selectedRepositoryStatus;
+        return !!status && !!status.branch && !status.upstreamBranch;
+    }
+    async publish(): Promise<void> {
+        const repository = this.repositoryTracker.selectedRepository;
+        const status = this.repositoryTracker.selectedRepositoryStatus;
+        const branch = status && status.branch;
+        if (!this.canPublish() || !repository || !branch) {
+            return;
+        }
+        const remote = await this.getRemote(repository, branch);
+        if (remote &&
+            await this.confirm('Publish changes', `This action will push commits to '${remote}/${branch}' and track it as an upstream branch.`)
+        ) {
+            try {
+                await this.push(repository, {
+                    remote, branch, setUpstream: true
+                });
+            } catch (e) {
+                this.error(e);
+            }
+        }
+    }
+    protected async getRemote(repository: Repository, branch: string): Promise<string | undefined> {
+        const remotes = await this.git.remote(repository);
+        if (remotes.length === 0) {
+            this.messageService.warn('Your repository has no remotes configured to publish to.');
+        }
+        return this.pick(`Pick a remote to publish the branch ${branch} to:`, remotes);
+    }
+
+    protected shouldPush(status: WorkingDirectoryStatus): boolean {
+        return status.aheadBehind && status.aheadBehind.ahead > 0 || true;
+    }
+    protected shouldPull(status: WorkingDirectoryStatus): boolean {
+        return status.aheadBehind && status.aheadBehind.behind > 0 || true;
+    }
+    protected async pull(repository: Repository, rebase: boolean): Promise<void> {
```

Comment:
Wouldn't it make sense to move this and `push` to `git`?

Reply:
sounds good. 

I also had to reimplement `push` below, since duigite `push` api does not allow to provide `setUpsteam` and `force` flags.  Do you think it should be moved to the backend as well?

Reply:
Whatever changes you need in dugite-extra, let me know, and I make those. We can do it in the context of a follow-up, whichever is more convenient for you.